### PR TITLE
add sanity tests for ParserWithLocation using RangeCheckerReducer

### PR DESCRIPTION
**Pull request by michaelficarra**
- `RangeCheckerReducer` ensures that subtree ranges do not overlap and are enclosed by parent node
- `copyNode` copies metadata from one node to another when we do transforms
- <s>fixed ordering of DataProperty parameters passed to reducers in Director</s> **edit:** I undid this for back-compat reasons

### CODE REVIEW

File: src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+
+		return outerBounds;
+	}
+
+	@NotNull
+	@Override
+	public RangeChecker reduceArrayBinding(
+		@NotNull ArrayBinding node, @NotNull ImmutableList<Maybe<RangeChecker>> elements,
+		@NotNull Maybe<RangeChecker> restElement
+	) {
```
Comment (disnet) [inline src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java:53]: Is this our style? Yuck.

File: src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+		return accept(node, super.reduceYieldGeneratorExpression(node, expression));
+	}
+
+	static class RangeChecker {
+		public final static Monoid<RangeChecker> MONOID = new Monoid<RangeChecker>() {
+			@NotNull
+			@Override
+			public RangeChecker identity() {
+				return new RangeChecker(Integer.MAX_VALUE, Integer.MIN_VALUE);
```
Comment (disnet) [inline src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java:690]: This seems backwards, or am I missing something?

File: src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java (hunk 1 above)
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java:53]: We don't really have a consistent style in this repo, and it was way better than what IntelliJ was giving me by default.

File: src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java (hunk 2 above)
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/shift/parser/RangeCheckerReducer.java:690]: It's the identity, so it should compare against any other range favourably. When compared "on the right", it should always start _after_ the end of the other range. Similarly, when compared "on the left", it should always end _before_ the beginning of the other range. So these numbers are right.

**Comment by disnet** [comment]
:+1:

**Pull request merged**